sourcepolicy/engine.go

```diff
@@ -0,0 +1,199 @@
+package sourcepolicy
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	"github.com/moby/buildkit/solver/pb"
+	spb "github.com/moby/buildkit/sourcepolicy/pb"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrSourceDenied is returned by the policy engine when a source is denied by the policy.
+	ErrSourceDenied = errors.New("source denied by policy")
+
+	// ErrTooManyOps is returned by the policy engine when there are too many converts for a single source op.
+	ErrTooManyOps = errors.New("too many operations")
+)
+
+// Engine is the source policy engine.
+// It is responsible for evaluating a source policy against a source operation.
+// Create one with `NewEngine`
+//
+// Rule matching is delegated to the `Matcher` interface.
+// Mutations are delegated to the `Mutater` interface.
+type Engine struct {
+	pol     *spb.Policy
```

Comment:
Maybe something for later, but we should support multiple policies here - while atm it's per job (I think?), we might also want to allow frontends to specify their own policy or have policy written in a global config somewhere?

I think we'd expect all the policies to be evaluated and come to the same conclusion, or fallthrough to a special default value (otherwise we have an error).

Reply:
Ideally, this field would also be added to gw `SolveRequest` https://github.com/moby/buildkit/blob/2e42ddce0f35e6a7a772a0f17da79cc35276764e/frontend/gateway/client/client.go#L120 . That would then be added to this array of policies with every nested level adding their policy in https://github.com/moby/buildkit/blob/2e42ddce0f35e6a7a772a0f17da79cc35276764e/solver/llbsolver/provenance.go#L144-L145 . If it makes it much more complicated we can leave it to follow-up, but it means that flows for building from a remote repository where policy is defined in that repository would not be possible atm.

Reply:
Added support for multiple policies and passthrough from the frontend.